### PR TITLE
Fix static checks failures for apiserver

### LINKED ISSUE #92402: fix staticcheck failures
Many of our packages are failing [staticcheck], we should fix all of them. See previously https://github.com/kubernetes/kubernetes/issues/36858 https://github.com/kubernetes/kubernetes/issues/90208 https://github.com/kubernetes/kubernetes/issues/81657

/help

To fix these, take a look at [hack/.staticcheck_failures] for a list of currently failing packages.

You can check a package by removing it from the list and running `make verify WHAT=staticcheck`.
Once the package is no longer failing, please file a PR including the fixes and removing it from the list.

Before filing your PR, you should run `hack/update-gofmt.sh` and `hack/update-bazel.sh` to auto format your code and update the build.

I recommend keeping PRs scoped down in size for review, fix a package or set of closely related packages, or a certain class of failures, to make it easier for your reviewers to handle. 

We don't need to fix all failures in one PR, but please avoid 100s of single character PRs, as it does cost time & resources to get each PR reviewed, tested, merged etc. :upside_down_face: 

I can help review _some_ of these, but many of them will require other reviewers that own the relevant code.

EDIT: I also don't recommend `/assign` ing to this one, many people are going to need to work on it and we want more people to join in.
You can link to this issue in your PR and comment here that you're working on certain packages to help avoid duplication.
EDIT2: Please do NOT put `Fixes #...` in your Pull Request, despite the template. That will close this PR. Instead put something like `Part of #92402`

I think each of these own code that is currently failing:
/sig testing
/sig storage
/sig api-machinery
/sig architecture
/sig cli
/sig instrumentation
/sig autoscaling
/sig cloud-provider

[staticcheck]: https://staticcheck.io/
[hack/.staticcheck_failures]: https://github.com/kubernetes/kubernetes/blob/master/hack/.staticcheck_failures

### PR DESCRIPTION
**Pull request by mmiranda96**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Address static check failures in the following directories:
- `staging/k8s.io/apiserver/pkg/endpoints`
- `staging/k8s.io/apiserver/pkg/server`

#### Which issue(s) this PR fixes:
Partially addresses #92402

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by pacoxu** [comment]
You may follow the[ suggestion from the CI ](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102899/pull-kubernetes-dependencies/1405086952784400384)to run `./hack/update-vendor.sh`.

File: staging/src/k8s.io/apiserver/pkg/server/routes/metrics.go [hunk 1]
```diff
@@ -30,7 +31,7 @@ type DefaultMetrics struct{}
 // Install adds the DefaultMetrics handler
 func (m DefaultMetrics) Install(c *mux.PathRecorderMux) {
 	register()
-	c.Handle("/metrics", legacyregistry.Handler())
+	c.Handle("/metrics", promhttp.Handler())
```
Comment (pacoxu) [inline staging/src/k8s.io/apiserver/pkg/server/routes/metrics.go:34]: This seems not to bre related with the static check. This change is in #81531.
And it caused the CI failure: https://github.com/kubernetes/kubernetes/pull/102899#issuecomment-862187749

Comment (mmiranda96) [reply staging/src/k8s.io/apiserver/pkg/server/routes/metrics.go:34]: This line gives the following check failure:
```
k8s.io/apiserver/pkg/server/routes/metrics.go:33:23: legacyregistry.Handler is deprecated: Please note the issues described in the doc comment of InstrumentHandler. You might want to consider using promhttp.Handler instead.  (SA1019)
```

I restored it to `legacyregister` and added a note.

**Comment by tiloso** [comment]
Hey @mmiranda96, please replace _fixes_ in the PR's description to avoid closing #92402 on merge. The code changes look good to me.

**Comment by mmiranda96** [comment]
Updated description, it shouldn't close the issue.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go [hunk 2]
```diff
@@ -229,6 +229,7 @@ type fancyResponseWriterDelegator struct {
 }
 
 func (f *fancyResponseWriterDelegator) CloseNotify() <-chan bool {
+	//lint:ignore SA1019 backwards compatibility
 	return f.ResponseWriter.(http.CloseNotifier).CloseNotify()
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:233]: Should we also mark `fancyResponseWriterDelegator.CloseNotify()` as deprecated?

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 3]
```diff
@@ -625,6 +626,7 @@ type fancyResponseWriterDelegator struct {
 }
 
 func (f *fancyResponseWriterDelegator) CloseNotify() <-chan bool {
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:628]: Mark this fancyResponseWriterDelegator.CloseNotify() as deprecated?

**Review by jpbetz — CHANGES_REQUESTED** [review]
Someone more familiar with certs is needed to review the NameToCertificate deprecation. I was able to review everything else.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go (hunk 2 above)
Comment (mmiranda96) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:234]: Done. Also: marking as deprecated makes the lint warning unnecessary.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go (hunk 3 above)
Comment (mmiranda96) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:630]: Done.

**Comment by mmiranda96** [comment]
Gentle ping for reviewers.
@deads2k
@dims

**Comment by BenTheElder** [comment]
/approve
for hack/

thanks for working on these!

**Comment by RA489** [comment]
@mmiranda96 can you please rebase this PR and squash the coomits to one?

**Comment by yuswift** [comment]
@mmiranda96 hi https://github.com/kubernetes/kubernetes/pull/102629 fixed vendor/k8s.io/apiserver/pkg/endpoints/filters. I think you need to rebase and fix the conflict.

**Comment by mmiranda96** [comment]
Closing this PR for a lack of reviewer.
/close

**Pull request closed without merging**